server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/AbstractEndpoint.java

```diff
@@ -39,50 +50,78 @@
    * The endpoint result for the ongoing transaction.
    */
   protected String my_endpoint_result;
-
+  
   /**
-   * Indicate that an internal server error has taken place.
+   * The ASM for this endpoint, load in the `before` of the endpoint
+   * and saved in the `after`.
    */
-  public void serverError(final Response the_response, final String the_log_message) {
-    the_response.status(HttpStatus.INTERNAL_SERVER_ERROR_500);
-    Main.LOGGER.error("server error 500: " + the_log_message);
-    my_endpoint_result = the_log_message;
+  protected AbstractStateMachine my_asm;
+  
+  /**
+   * Halts the endpoint execution by ending the request and returning the
+   * most recently set response code and endpoint result.
+   */
+  protected final void halt(final Response the_response) {
+    Spark.halt(the_response.status(), my_endpoint_result);
   }
 
   /**
-   * Indicate that some data was not found.
+   * @return The classname of the ASM used by this endpoint.
    */
-  public void dataNotFound(final Response the_response, 
-                           final String the_log_message) {
-    the_response.status(HttpStatus.NOT_FOUND_404);
-    Main.LOGGER.error("server error 404: " + the_log_message);
-    my_endpoint_result = the_log_message;
+  protected abstract Class<? extends AbstractStateMachine> asmClass(); 
+  
+  /**
+   * Load the appropriate ASM from the database and make sure that
+   * the transition we wish to take is legal.
+   */
+  protected void loadAndCheckASM(final Response the_response) {
+    // get the state of the DoS dashboard
+    try {
+      my_asm = (AbstractStateMachine) (asmClass().getConstructor().newInstance());
+    } catch (final Exception e) {
+      // something went horribly wrong
+    }
+    final PersistentASMState state = 
+        PersistentASMStateQueries.get(asmClass(), null);
+    state.applyTo(my_asm);
+    // check that we are in the right ASM state
+    if (!my_asm.enabledASMEvents().contains(endpointEvent())) {
+      illegalTransition(the_response,
+                        endpointName() + 
+                        "attempted to apply illegal event " + endpointEvent() + 
+                        " from state " + my_asm.currentState());
+    }
   }
-
+  
   /**
-   * Indicate that data that the client provided is ill-formed.
+   * Save the ASM back to the database.
    */
-  public void badDataContents(final Response the_response, final String the_log_message) {
-    the_response.status(HttpStatus.UNPROCESSABLE_ENTITY_422);
-    Main.LOGGER.error("bad data from client 422: " + the_log_message);
-    my_endpoint_result = the_log_message;
+  protected void transitionAndSaveASM()  {
+    my_asm.stepEvent(endpointEvent());
+    final PersistentASMState new_state = PersistentASMState.stateFor(my_asm);
```

Comment:
This needs to take the actual persistent state object (either because we save it somewhere, or because we ask for it again) returned by `PersistentASMStateQueries`and call `updateFrom()` on it. This new object will have a different database key, and when you save it, there will then be two different states for the same ASM (and the system will understandably get upset next time it tries to load one).